Source/Core/VideoCommon/LightingShaderGen.h

```diff
@@ -62,7 +62,8 @@ static void GenerateLightShader(T& object, LightingUidData& uid_data, int index,
 		case LIGHTATTN_NONE:
 		case LIGHTATTN_DIR:
 			object.Write("ldir = normalize(" LIGHT_POS".xyz - pos.xyz);\n", LIGHT_POS_PARAMS(index));
-			object.Write("attn = 1.0f;\n");
+			object.Write("attn = 1.0;\n");
+			object.Write("if (length(ldir) == 0.0)\n\t ldir = float3(1.0, 1.0, 1.0);\n");
```

Comment:
The problem isn't that you're checking length(ldir) instead of the individual dimensions; actually length(ldir) is the thing you want to check given how floating-point math works. The problem is that you have to perform the check before you normalize ldir: normalizing a vector is just "vec / length(vec)", which gives unpredictable results if length(vec) == 0.  In practice, it might give usable results on graphics hardware, but the software backend's implementation of Normalized() will return (NaN, NaN, NaN) for a zero input.